java-checks/src/main/java/org/sonar/java/checks/ValueBasedObjectsShouldNotBeSerializedCheck.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ValueBasedUtils;
+import org.sonar.java.checks.serialization.SerializableContract;
+import org.sonar.java.model.ModifiersUtils;
+import org.sonar.java.resolve.ParametrizedTypeJavaType;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.Modifier;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S3437")
+public class ValueBasedObjectsShouldNotBeSerializedCheck extends BaseTreeVisitor implements JavaFileScanner {
+  
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+  
+  @Override
+  public void visitClass(ClassTree clas) {
+    if (isSerializable(clas) && !SerializableContract.hasSpecialHandlingSerializationMethods(clas)) {
+      clas.members().stream()
+        .filter(member -> member.is(Tree.Kind.VARIABLE))
+        .filter(member -> !isStatic((VariableTree) member))
+        .filter(member -> !isTransient((VariableTree) member))
+        .filter(member -> isSerializableAndValueBased(((VariableTree) member).type().symbolType()))
+        .forEach(member -> context.reportIssue(
+          this,
+          ((VariableTree) member).simpleName(),
+          "Make this value-based field transient so it is not included in the serialization of this class."));
+    }
+
+    super.visitClass(clas);
+  }
+
+  protected static boolean isSerializableAndValueBased(Type t) {
+    // we check first the ParametrizedTypeJavaType, in order to filter out the non-serializable
+    // generic value-based class Optional<T>
+    if (t instanceof ParametrizedTypeJavaType) {
```

Comment:
Here would be better to downcast (safely) the variable to a `JavaType` (only implementation of API) and use the isParameterized method.